### PR TITLE
Added accessory button to message cell

### PR DESCRIPTION
**Pull request by adubr**
## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have followed the [coding style](https://github.com/jessesquires/HowToContribute#style-guidelines), and reviewed the [contributing guidelines](https://github.com/jessesquires/JSQMessagesViewController/blob/develop/.github/CONTRIBUTING.md). Confirmation: 💪😎👊

## What's in this pull request?

Added an accessory button to message collection view cells.

- A `UIButton` added to the message view via XIBs
- A dedicated delegate/protocol for the callback is implemented
- A class for configuring accessory button appearance is provided. A client can configure an image of the button. The button has 4 modes of visibility:
  - Hidden
  - Visible for media messages only
  - Visible for text messages only
  - Visible at all times
- A setting that enables the accessory button for media messages is added to the demo project
- All changes and additions are documented
- Small fix in tests

### CODE REVIEW

File: JSQMessagesViewController/Model/JSQMessagesViewAccessoryButtonConfiguration.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+//
+//
+//  GitHub
+//  https://github.com/jessesquires/JSQMessagesViewController
+//
+//
+//  License
+//  Copyright (c) 2014 Jesse Squires
+//  Released under an MIT license: http://opensource.org/licenses/MIT
+//
```
Comment (jessesquires) [inline JSQMessagesViewController/Model/JSQMessagesViewAccessoryButtonConfiguration.h]: for all new files, let's keep the comment headers consistent with the rest of the project 😄

**Comment by jessesquires** [comment]
Thanks so much @adubr ! 😄  Overall, this looks great. Nice work! 🎉 

We'll definitely merge this after a few changes.

Minor:
- Keep comment headers consistent for new files
- Need to rebase this branch on `develop`

Design:

Let's remove `JSQMessagesViewAccessoryButtonConfiguration` completely. While I generally like this kind of pattern, I think in this instance it makes the API too complex. We can simplify:

- We can setup the cell with the default button and image. Because the button is public, clients can customize the button easily in `cellForItemAtIndexPath:`
- Rather than `NS_OPTIONS` for the mode, we can provide a delegate method, `shouldShowAccessoryButtonAtIndexPath:`. This is simpler and more flexible for clients. There's not a great place for this new method, so for now, let's just add it to the existing `dataSource`. I'll be refactoring this later.

**Comment by adubr** [comment]
@jessesquires thanks for the review! 🤓

No worries, I'll do this in couple of days.

Though it seems to me kind of inconsistent to have two totally different places for the same task of configuring the button's appearance.

**Comment by jessesquires** [comment]
> Though it seems to me kind of inconsistent to have two totally different places for the same task of configuring the button's appearance.

Well, it's more like:

(1) You get a default accessory button. Then you implement `shouldShow...` to hide/show it.

OR

(2) The above, but you chose to customize the button in `cellForItemAtIndexPath:`

---------------------------------------

Actually. Now that I think about it -- you *could* just set `.hidden` from `cellForItemAtIndexPath:`. So maybe that extra delegate method isn't worth it? 

Maybe we just provide this API with defaults and let clients do what they want from `cellForItemAtIndexPath:`.

Thoughts? 

> /cc: @eliburke @sebastianludwig @MacMeDan @GianniCarlo

**Comment by MacMeDan** [comment]
I think that is definitely a better experience when there is already a default implementation with the ability to override for customization.  Also by default I look to implementing my own customizations in `cellForItemAtIndexPath:` So I would say it makes logical since but I have not looked thought everything on this pull request but initially those are my thoughts.

**Comment by adubr** [comment]
> Now that I think about it -- you could just set .hidden from cellForItemAtIndexPath:. So maybe that extra delegate method isn't worth it?

That's what I'm talking about!

I think this issue points to a bigger one (which you're certainly aware of). There are two design patterns for enhancing functionality of a class: inheritance and delegation. The latter being less fragile and less error prone than the former. It also provides specific points of customization.

If you take UIKit examples like `UICollectionViewController` or `UITableViewController` – they are merely convenience classes, you don't have to subclass them (and in my experience you rarely do), while all the work is done by the corresponding views and their delegates.

In case of `JSQMessagesViewController` inheritance and delegation are **mixed up**. If a user is not familiar with the architecture yet, how would she/he know where to look for a particular customization point?

This leads to some weird constructions like

```Objective-C
cell.cellTopLabel.attributedText =
          [collectionView.dataSource collectionView:collectionView
           attributedTextForCellTopLabelAtIndexPath:indexPath];
```
— we're in a data source using _(what it looks like in the code)_ a data source of **another** object (but in fact it's just `self`)! This actually means that you can't get away with delegation alone (which is not only a common expectation but kind of a purpose of delegation), you must subclass.

Or:

<pre>
- (id<JSQMessageBubbleImageDataSource>)collectionView:(JSQMessagesCollectionView *)collectionView messageBubbleImageDataForItemAtIndexPath:(NSIndexPath *)indexPath
{
    <b>JSQMessage *message = [self.demoData.messages objectAtIndex:indexPath.item];</b>
    
    if ([message.senderId isEqualToString:self.senderId]) {
        return self.demoData.outgoingBubbleImageData;
    }
    
    return self.demoData.incomingBubbleImageData;
}
</pre>
— wait a sec, aren't we supposed to use a designated and nicely documented data source method to fetch a message?

I see three possible solutions
- A user either have to subclass and override methods like `cellForItemAtIndexPath:` (with **no** `collectionView:` prefix) and doesn't have to mess with all these tangled delegate calls.
- Or she/he implements delegates and data sources, and doesn't have to subclass `JSQMessagesViewController` (other bits of its functionality can be moved to other classes, e.g. some container views).
- Last but not least, since Apple allows us to build hierarchies of view controllers, the `JSQMessagesViewController` class can be made **closed for subclassing** too, but customizable via various "external" delegates.

**Comment by eliburke** [comment]
Uh oh design patterns. This is starting to feel like being at work. 😄
I am starting to agree with some of @jessesquires views on enums and configurability. With the general complexity of this framework and the number of people who use it without reading any of the documentation, I think we need to aim for simplicity. 

This is what I would vote for, if this were up for a vote: 
A delegate method for beginners, and inheritance for more advanced users. First, a simple BOOL delegate like `shouldShowAccessoryButtonAtIndexPath`. This is easy to find if you look at the headers or demo project, and simple to understand. Second, instead of customizing via `celForItemAtIndexPath`, add a new method `accessoryViewForItemAtIndexPath` and give it a default implementation that returns the share button you have created. Advanced users can subclass and override if they want something beyond the share button. 

This accomplishes the initial goal of a simple accessory button, while also providing a single powerful customization point 

Some other thoughts...
1) I'm personally in favor of fewer protocols rather than having a lot of small ones (e.g. JSQMessagesViewAccessoryButtonDelegate). 
2) Although the default behavior is a button, and we want a built-in tap delegate, I really think it should be implemented as vanilla UIView. 
3) the semantics of *showing* a button should not the be same as setting it hidden. If a button is *shown* there should be a property that can be set hidden=YES/NO. If it is not shown, no view should be allocated.

**Comment by jessesquires** [comment]
Hey all 😊 thanks for the feedback.

@adubr 
> In case of JSQMessagesViewController inheritance and delegation are mixed up.

Yes I agree, and this is because of legacy reasons. However, the pattern really isn't much different than `UICollectionViewController` where `self` *is* the `delegate` and `dataSource`. The decision to build this in a similar way to `UICollectionViewController` was made long ago. 😅  

@eliburke 
Lots of good points there 👍 

----------------------

My goal with `v8.0` is to address all of these points. In particular, we'll have a separate DataSource class and a CellFactory class to move all of this logic out of the view controller. I want it to be possible to use this library *without* having to subclass the view controller — that is, if you use the defaults, or do minor customization, then you wouldn't need to subclass. However, subclassing would still be possible if a user chooses to do that. #453 is mostly tracking this, though I'm waiting for this PR to merge before I start tearing things up. 😄 

Having said that, let's move forward with this PR using the `cellForItemAtIndexPath:` approach (no extra delegates) — for now. After we merge and after I push initial changes for #453 we can iterate on this. Sound good to everyone?

**Comment by jessesquires** [comment]
✨ Friendly ping ✨

**Comment by adubr** [comment]
Hi all

sorry for the delay.
@jessesquires please check updated commit

**Comment by jessesquires** [comment]
Thanks @adubr ! 🎉

**Comment by apwelsh** [comment]
@eliburke Your comment about people just using the library and not reading the document is perplexing.  I am an avid reader of documentation, and for the life of me, I cannot find any documentation beyond.  Everything I know about this library is by reverse-engineering the source, and searching the web.  I am still somewhat lost how best to implement buttons on a text message, for a few reasons.

a) Using a media item, I have no way of determining the maximum width for my messages.
b) I cannot use the existing messaging techniques for the table view touches, since there is no apparent technique documented anywhere to send touch events for when each button is pressed.
c) Subclassing (extending) the default messaging cells to add buttons is complicated, can confusing, because without documentation, I am at a loss for the best way to extend the cell, to render the buttons.  I would like to think I could modify the current cells, by injecting my own buttons into the view, but again, with a lack of documentation, I am at a loss in determining if this is even possible, since I cannot easily determine which bindings (like constraints, and controls) are bound where.

so, if the lack of documentation forces us to try many techniques to solve our challenges.  Your comment suggests there is documentation.  Where?  I would love to read documentation that explains the control's features, and implementation options.

**Comment by eliburke** [comment]
@apwelsh kinda strange to take offense at one sentence in a 4 month old message that didn't involve you, but ok. 

First off, I said "given the number of people...", not "given that nobody...". If you read docs, good for you. 

Second, if you don't consider this "docs" you must not spend much time on github: http://cocoadocs.org/docsets/JSQMessagesViewController/7.3.4/

There is a lengthy README, a FAQ, a huge set of API docs, and a well commented demo. And if you think the devs of every one of the 10000+ projects using JSQ took the time to read or search what is available before opening an issue, take a few hours to review some of the 1900 issues that have been opened. 

I'm sorry that you can't find easily find docs or an issue that addresses your particular problem. Try to remember this project is maintained by volunteers. Time spent answering issues, creating new features, writing docs, and anticipating EVERY possible app's use case is seldom rewarded in any way.

It would be nice if more people "gave back" by explaining what they did or contributing their solutions, but it's a very very small minority that takes the time to do so. Keep that in mind as you work through your problems. And if you have a more specific question, feel free to open a new issue! 😀

**Comment by apwelsh** [comment]
@eliburke I am not complaining.  And I do try to give back.  I am very happy that this project is actually being maintained, so many get abandoned.  
But when it comes to documentation, what is missing is an explanation of how to customize the system.  Most of it ends with vague explanation that it is possible, and nothing more.  In this thread the OP was discussing a pull request to address the accessory button, and the discussion took a turn into design patterns.  I was offended by the claim about failing to read the docs, because I have read everything I can find, and there are deficiencies, many deficiencies.   Which is why there are so many issues like these.  
Rest assured, as I learn to implement the things I need, I will be enhancing the documentation in GtHub with my own pull requests.  My challenges are not unique, nor one-off, as I have found many people asking the questions I am looking for, and all of them are based on the knowledge the existing documentation provides.   
Sure, there are a log of issue that deal with the basics, that are clearly documented, but when someone addresses issues that clearly are advanced, and not part of the documentation, I was frustrated.

**Comment by jessesquires** [comment]
To be clear -- **advanced** cell customization is not really advertised as a first-class feature. It is *possible* but doesn't have first-class support. We'll be trying to address this pain point in v8.0 of the library, as it has been frequently discussed and requested.

**Pull request merged**